gax/src/test/java/com/google/api/gax/retrying/ContextAwareScheduledRetryingExecutorTest.java

```diff
@@ -0,0 +1,366 @@
+              .build();
+      RetryingExecutorWithContext<String> executor =
+          getRetryingExecutor(getAlgorithm(FAILING_RETRY_SETTINGS, 0, null), localExecutor);
+      RetryingFuture<String> future =
+          executor.createFuture(
+              callable,
+              FakeCallContext.createDefault().withTracer(tracer).withRetrySettings(retrySettings));
+      future.setAttemptFuture(executor.submit(future));
+
+      Thread.sleep(30L);
+
+      boolean res = future.cancel(false);
+      assertTrue(res);
+      assertFutureCancel(future);
+      assertTrue(future.getAttemptSettings().getAttemptCount() < 4);
+      localExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testCancelIsTraced() throws Exception {
+    ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+    FailingCallable callable = new FailingCallable(4, "SUCCESS", tracer);
+    RetrySettings retrySettings =
+        FAST_RETRY_SETTINGS
+            .toBuilder()
+            .setInitialRetryDelay(Duration.ofMillis(1_000L))
+            .setMaxRetryDelay(Duration.ofMillis(1_000L))
+            .setTotalTimeout(Duration.ofMillis(10_0000L))
+            .build();
+    RetryingExecutorWithContext<String> executor =
+        getRetryingExecutor(getAlgorithm(FAILING_RETRY_SETTINGS, 0, null), localExecutor);
+    RetryingFuture<String> future =
+        executor.createFuture(
+            callable,
+            FakeCallContext.createDefault().withTracer(tracer).withRetrySettings(retrySettings));
+    future.setAttemptFuture(executor.submit(future));
+
+    Thread.sleep(30L);
+
+    boolean res = future.cancel(false);
+    assertTrue(res);
+    assertFutureCancel(future);
+
+    Mockito.verify(tracer).attemptCancelled();
+    localExecutor.shutdownNow();
+  }
+
+  @Test
+  public void testCancelProxiedFutureAfterStart() throws Exception {
+    // this is a heavy test, which takes a lot of time, so only few executions.
+    for (int executionsCount = 0; executionsCount < 2; executionsCount++) {
+      ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+      FailingCallable callable = new FailingCallable(5, "SUCCESS", tracer);
+      RetrySettings retrySettings =
+          FAST_RETRY_SETTINGS
+              .toBuilder()
+              .setInitialRetryDelay(Duration.ofMillis(1_000L))
+              .setMaxRetryDelay(Duration.ofMillis(1_000L))
+              .setTotalTimeout(Duration.ofMillis(10_0000L))
+              .build();
+      RetryingExecutorWithContext<String> executor =
+          getRetryingExecutor(getAlgorithm(FAILING_RETRY_SETTINGS, 0, null), localExecutor);
+      RetryingFuture<String> future =
+          executor.createFuture(
+              callable,
+              FakeCallContext.createDefault().withTracer(tracer).withRetrySettings(retrySettings));
+      future.setAttemptFuture(executor.submit(future));
+
+      Thread.sleep(50L);
```

Comment:
DANGER! This adds 50 milliseconds to every test run.

Reply:
Same as above.